### DISCUSSION TITLE
Billboards

### DISCUSSION THREAD

**Discussion by alienatorZ** [iTowns/itowns#2064] [Q&A] [answered]
Is there an example for how to create billboards similar to billboards in Cesium?

Thanks!

**Comment by alienatorZ** [comment]
Ok, I think that I figured that out by using the LabelLayer example. Now my question is: How do I move a label in real time? Imagine a label representing a persons position. How is the best way to update the position?

**Comment by jailln** [comment]
I would advise to use threejs directly to do so instead of `LabelLayer` that has been designed to display text labels with static positions.
With threejs, you can use a [THREE.Points](https://threejs.org/docs/index.html?q=points#api/en/objects/Points) object and add a texture in the [PointsMaterial's map attribute](https://threejs.org/docs/index.html?q=points#api/en/materials/PointsMaterial.map) (that is the image of your billboard). You can get inspiration from the following threejs example: https://threejs.org/examples/?q=bill#webgl_points_billboards
Once you created your `THREE.Points` object, you can add it to itowns scene with `view.scene.add(points);`. 
To update your billboards positions, you then only have to modify the [THREE.Points geometry](https://threejs.org/docs/index.html?q=points#api/en/objects/Points.geometry) which stores the points positions.
Let us know if that works out.

**Reply by alienatorZ** [reply]
Oh that seems like a better solution. I will give that a try.

**Reply by alienatorZ** [reply]
I tried this, but I dont see anything shown. 


    let material;
    const sprite = new THREE.TextureLoader().load( 'icon.png' );
    var cameraTargetPosition = view.controls.getLookAtCoordinate();
    const geometry = new THREE.BufferGeometry();
    geometry.setAttribute( 'position', cameraTargetPosition );
    material = new THREE.PointsMaterial( { size: 35, sizeAttenuation: false, map: sprite, alphaTest: 0.5, transparent: false } );
    material.color.setHSL( 1.0, 0.3, 0.7 );
    const particles = new THREE.Points( geometry, material );
    view.scene.add( particles );


I verified the icon is available and being fetched. Any ideas?

**Reply by alienatorZ** [reply]
I also recieve this error when zooming wayy out:

three.cjs:8289 Uncaught TypeError: attribute.onUploadCallback is not a function
    at createBuffer (three.cjs:8289:13)
    at Object.update2 [as update] (three.cjs:8375:27)
    at Object.update2 [as update] (three.cjs:11061:15)
    at Object.update2 [as update] (three.cjs:11417:15)
    at projectObject (three.cjs:18179:31)
    at projectObject (three.cjs:18198:4)
    at WebGLRenderer.render (three.cjs:18074:3)
    at c3DEngine2.<anonymous> (c3DEngine.js:86:21)
    at MainLoop2._renderView2 (MainLoop.js:365:20)
    at MainLoop2._step5 (MainLoop.js:349:61)

**Reply by jailln** [reply]
I think it's because you're adding a `Coordinates` object  as a buffer attribute in the line `geometry.setAttribute( 'position', cameraTargetPosition );` instead of a [BufferAttribute](https://threejs.org/docs/index.html?q=buffergeo#api/en/core/BufferAttribute).

Try something like this instead: (I haven't tested it so you might need some adaptations)
```
const posArr = cameraTargetPosition.toVector3().toArray();
const posTypedArr = new Float32Array(posArr);
const posAttrib = new THREE.BufferAttribute(posTypedArr, 3);
geometry.setAttribute( 'position', posAttrib );
```

**Reply by alienatorZ** [reply]
Thank you. It eliminated the zooming error but I still cannot see anything on the map.

**Reply by alienatorZ** [reply]
This is my complete function:



import * as THREE from 'three';
export function CreateSVGObj(view){
    let material;
    const sprite = new THREE.TextureLoader().load( 'icon.png' );
    var cameraTargetPosition = view.controls.getLookAtCoordinate();
    cameraTargetPosition.altitude += 30;
    const geometry = new THREE.BufferGeometry();
    const posArr = cameraTargetPosition.toVector3().toArray();
    const posTypedArr = new Float32Array(posArr);
    const posAttrib = new THREE.BufferAttribute(posTypedArr, 3);
    geometry.setAttribute( 'position', posAttrib );
    material = new THREE.PointsMaterial( { size: 135, sizeAttenuation: false, map: sprite, alphaTest: 0.5, transparent: false } );
    material.color.setHSL( 1.0, 0.3, 0.7 );
    const particles = new THREE.Points( geometry, material );
    view.scene.add( particles );
    view.notifyChange(view.camera.camera3D, true);
}

**Reply by jailln** [reply, marked as answer]
Ok it's because `var cameraTargetPosition = view.controls.getLookAtCoordinate();` returns coordinates in `EPSG:4326` (lat,long, alt) and you need to place the point in the coordinates of the view so you need an intermediary coordinates conversions in the view crs before setting the point position: `var posGeo = cameraTargetPosition.as(view.referenceCrs);`

**Reply by alienatorZ** [reply]
That worked!

**Reply by alienatorZ** [reply]
Thank you!
